### PR TITLE
Implement Clustering for Snowflake

### PR DESCRIPTION
**Pull request by bastienboutonnet**
## Aim
Brings clustering to snowflake.

Table materialisations will leverage an order by via wrapping the SQL in a `select * from ( {{sql}} ) order by ( {{cluster_by_keys}}`
Incremental materialisation will create table as above and followed by an alter statement `alter table {{relation}} cluster by ({{cluster_by_keys}})` to leverage Snowflake's automatic clustering.
This approach was discussed in #634

This is a duplicate of PR #1591  that I had to re open due to killing my fork in the meantime... All discussion can be found there though.

### CODE REVIEW

**Comment by bastienboutonnet** [comment]
Btw @drewbanin CLA is now signed...

**Comment by drewbanin** [comment]
Thanks @bastienboutonnet! Kicking off the tests now :)

@cla-bot check

File: plugins/snowflake/dbt/include/snowflake/macros/adapters.sql [hunk 1]
```diff
@@ -1,14 +1,33 @@
 {% macro snowflake__create_table_as(temporary, relation, sql) -%}
   {%- set transient = config.get('transient', default=true) -%}
+  {%- set cluster_by_keys = config.get('cluster_by', default=none) -%}
+  {%- set enable_automatic_clustering = config.get('automatic_clustering', default=false) -%}
+  {%- if cluster_by_keys is not none and cluster_by_keys is string -%}
+    {%-  set cluster_by_keys = [cluster_by_keys] -%}
+  {%- endif -%}
+  {%- set cluster_by_string = cluster_by_keys|join(", ")-%}
```
Comment (drewbanin) [inline plugins/snowflake/dbt/include/snowflake/macros/adapters.sql]: I think this tries to call `join` on `none` when the `cluster_by` config is unset. Can we try something like:
```
{%- if cluster_by_keys is not none and cluster_by_keys is string -%}
    {%-  set cluster_by_keys = [cluster_by_keys] -%}
    {%- set cluster_by_string = cluster_by_keys|join(", ")-%}
{% else %}
    {%-  set cluster_by_keys = [] -%}
    {%- set cluster_by_string = ""-%}
{%- endif -%}

Comment (drewbanin) [reply plugins/snowflake/dbt/include/snowflake/macros/adapters.sql]: and then below, we'll want to do:
```
{% if cluster_by_string is not none -%}
    alter table {{relation}} cluster by ({{cluster_by_string}});
{%- endif -%}
```

Comment (bastienboutonnet) [reply plugins/snowflake/dbt/include/snowflake/macros/adapters.sql]: Oh damn, that was really stupid of me... 

But actually the whole logic is pretty wild here... I was doing really stupid stuff but it should work after my last commit. Let me know if you don't agree with that flow. I tested it and it works both for list of strings or just one string.

I also noticed a corner case. We're allowing `alter table {{relation}} resume recluster;` to run on the condition that this config argument be set to true. The issue with that is that if the table was not created as a clustered table (i.e. that no `cluster_by_keys` were provided) then Snowflake will throw an error that the table is not clustered. 

I changed this piece of logic as well. Let me know what you think.

File: plugins/snowflake/dbt/include/snowflake/macros/adapters.sql [hunk 2]
```diff
@@ -1,14 +1,37 @@
...
+      as (
+        {%- if cluster_by_string is not none -%}
+          select * from(
+            {{ sql }}
+            ) order by ({{ cluster_by_string }})
+        {%- else -%}
+          {{ sql }}
+        {%- endif %}
+      );
+    {% if cluster_by_string is not none -%}
```
Comment (drewbanin) [inline plugins/snowflake/dbt/include/snowflake/macros/adapters.sql]:
```suggestion
    {% if cluster_by_string is not none and not temporary -%}
```

We probably don't want to cluster temporary tables - those are just going to be used to stage data that we want to insert into a permanent table.

File: plugins/snowflake/dbt/include/snowflake/macros/adapters.sql [hunk 3]
```diff
@@ -1,14 +1,37 @@
...
+      alter table {{relation}} cluster by ({{cluster_by_string}});
+    {%- endif -%}
+    {% if enable_automatic_clustering and cluster_by_string is not none  -%}
```
Comment (drewbanin) [inline plugins/snowflake/dbt/include/snowflake/macros/adapters.sql]:
```suggestion
    {% if enable_automatic_clustering and cluster_by_string is not none and not temporary  -%}
```

Same here - i don't think we'll want to resume reclustering on a temp table!

**Comment by drewbanin** [comment]
Merging this - thanks for your hard work in this contribution @bastienboutonnet!!

**Comment by bastienboutonnet** [comment]
pleasure as always!

**Pull request merged**